studio/components/LanguageSelector.tsx

```diff
@@ -0,0 +1,127 @@
+import { Box, Button, Card, Checkbox, Flex, useTheme } from "@sanity/ui";
+import React from "react";
+import { ArrayOfObjectsInputProps, PatchEvent, set } from "sanity";
+
+import {
+  Language,
+  supportedLanguages,
+} from "internationalization/supportedLanguages";
+
+const colorMap = {
+  dark: {
+    selected: "#29356a",
+    default: "rgba(19, 20, 27, 0.5)",
+  },
+  light: {
+    selected: "#E3F2FD",
+    default: "white",
+  },
+};
+
+const LanguageSelector = (props: ArrayOfObjectsInputProps<Language>) => {
+  const { value = [], onChange } = props;
+  const theme = useTheme();
+  const prefersDark = theme.sanity.v2?.color._dark ?? false;
+  const themeType = prefersDark ? "dark" : "light";
+  const currentDefaultLanguage: string | null =
+    value.find((lang) => lang.default)?.id || null;
+
+  const getNextDefaultLanguage = (updatedValue: Language[], lang: Language) => {
+    const indexInSupported = supportedLanguages.findIndex(
+      (language) => language.id === lang.id,
+    );
+    return (
+      supportedLanguages
+        .slice(indexInSupported + 1)
+        .find((language) =>
+          updatedValue.some((item) => item.id === language.id),
+        ) ||
+      supportedLanguages.find((language) =>
+        updatedValue.some((item) => item.id === language.id),
+      )
+    );
+  };
+
+  const handleLanguageSelection = (lang: Language, isSelected: boolean) => {
+    if (isSelected && value.length === 1) return;
+
+    let updatedValue = isSelected
+      ? value.filter((item) => item.id !== lang.id)
+      : [...value, { ...lang, default: false }];
+
+    if (isSelected && lang.id === currentDefaultLanguage) {
+      const nextDefaultLanguage = getNextDefaultLanguage(updatedValue, lang);
+      updatedValue = updatedValue.map((item) => ({
+        ...item,
+        default: item.id === (nextDefaultLanguage?.id || null),
+      }));
+    }
+
+    onChange(PatchEvent.from(set(updatedValue)));
+  };
```

Comment:
Just a suggestion for a cleaner iterator for finding a new default, and an abort if a new default could not be found (edge-case that should not really happen).
```suggestion
const getNextDefaultLanguageId = (
    updatedValue: Language[],
    oldDefaultLanguage: Language,
  ) => {
    const oldIndex = supportedLanguages.findIndex(
      (language) => language.id === oldDefaultLanguage.id,
    );
    for (let i = 0; i < supportedLanguages.length; i++) {
      const languageId =
        supportedLanguages[(i + oldIndex) % supportedLanguages.length].id;
      if (updatedValue.some((item) => item.id === languageId)) {
        return languageId;
      }
    }
    return null;
  };

  const handleLanguageSelection = (lang: Language, isSelected: boolean) => {
    if (isSelected && value.length === 1) return;

    let updatedValue = isSelected
      ? value.filter((item) => item.id !== lang.id)
      : [...value, { ...lang, default: false }];

    if (isSelected && lang.id === currentDefaultLanguage) {
      const nextDefaultLanguageId = getNextDefaultLanguageId(
        updatedValue,
        lang,
      );
      if (nextDefaultLanguageId === null) {
        // abort selection change
        return;
      }
      updatedValue = updatedValue.map((item) => ({
        ...item,
        default: item.id === nextDefaultLanguageId,
      }));
    }

    onChange(PatchEvent.from(set(updatedValue)));
  };
```